### ISSUE TITLE
how to build rusqlite on windows?

### ISSUE THREAD

**Issue by majkcramer** [jgallagher/rusqlite#46]
On Linux eco system i had no problems. But on windows I feel a little bit helpless. Could you provide some short hints on how to build rusqlite on windows, please?

Install pkg_config for windows? 
Which files I have to download from sqlite hompage? 
Where to place? 
...

Any hints?

**Comment by jgallagher** [comment]
Sorry, I don't have access to a Windows box for development at the moment. I know Rust's support on Windows is evolving too. Have you tried asking in the #rust IRC channel? There might be folks there who can help.

**Comment by jgallagher** [comment]
Oh! If you do figure something out, please reply; I'd love to document this.

**Comment by majkcramer** [comment]
I have no final solution for now. Only figured out something. Some steps must be wrong.

## Part #a: compile sqlite.dll.a
- Download mingw32 or mingw64 
- Download the sqlite amalgamation source code from https://www.sqlite.org/download.html
- Install mingw
- Decompress amalgamation source
- "Start Terminal" from mingw Startmenu entry (cmd.exe with env path to mingw/bin/ folder). Do the following 2 steps within the cmd Terminal
- cd to you decompresses amalgamation source
- "dlltool --def sqlite3.def --dllname sqlite3.dll --output-lib sqlite3.a"
- now there should be a sqlite3.a within that folder
- optional: check the sqlite.a does not contain debugger symbols: "nm --debug-syms sqlite.a" (For debugger symbols the second column indicates N.)

## Part #b: download and setup pkg-config
Hint: Do yourself a favor und download the gtk+ bundle instead only "pkg-config" and trying to fulfill required dependency
- Download gtk+ bundle from http://www.gtk.org/download/index.php 
- Decompress the gtk+ bundle
- Copy "sqlite3.def" from amalgamtion source to {gtk+ folder}/lib/
- Copy your compileted "sqlite3.a" to {gtk+ folder}/lib/ as "sqlite3.dll.a"
- Within {gtk+ folder}/lib/pkgconfig/ create a "sqlite3.pc" file with following contents:
```
# Package Information for pkg-config

prefix=/srv/win32builder/fixed_364/build/win32
exec_prefix=${prefix}
libdir=${exec_prefix}/lib
includedir=${prefix}/include

Name: sqlite3
Description: SQL database engine lib
Version: 3.7.7.1
Libs: -L${libdir} -lsqlite3
Libs.private: -ldl -lpthread 
Cflags: -I${includedir}
```
- Open cmd shell and cd to {gtk+ folder}/bin/ folder and type "pkc-config --list-all". The list should contain you "sqlite3" entry.

## Part #c: build rusqlite
- Open your Rust Shell
- Append your mingw/bin/ folder to env path (for example: 'set PATH=%PATH%;"C:\Program Files\mingw-w64\x86_64-4.9.2-posix-seh-rt_v4-rev2\mingw64\bin"') to prevent yourself from 'bug' https://github.com/rust-lang/time/issues/78
- Append your gtk+ bundle /bin folder to env path
- cargo build
-> should compile without errors now

## Part d#: test rusqlite
- Download sqlite precompiled dll from sqlite.org
- Copy dll to rusqlite/target/debug/ Folder
- cargo test
-> should compile without errors now
-> BUT I get a "application error" (0xc00007b) wich is mainly generate because of mixing up 32bit environment with 64bit one. It means for example you have a 32bit application that attempts to load a 64bit dll.

## Final thoughts
My Env is 64 Bit Windows 8.1. I used MingW64. And downloaded 32-Bit sqlite.dll on Part d#. To be sure my final error is due to mix 64Bit sqlite.a and 32Bit sqlite.dll i should compile 64 Bit for myself (dont know how yet) or download anywhere else (have not found). So far ... 

Thanks for your patience about my bad english ;-)

**Comment by photex** [comment]
I was able to get this going just using msys2.

Using pacman I installed the following: `mingw64/mingw-w64-x86_64-pkg-config` and `mingw64/mingw-w64-x86_64-sqlite3`

Also generally speaking, I've had to be sure I've installed GCC in order to actually compile most foo-sys packages: `mingw64/mingw-w64-x86_64-gcc`

For those who may not know, you install packages with pacman like so: `pacman -Ss <package name>`
And you can update everything using `pacman -Syu`.

I'm just using powershell (for no particular reason honestly) but I've added the relevant msys and mingw paths to my system environment. On my system it's the following:
`C:\msys64\mingw64\bin;c:\msys64\usr\bin`
The order here is important because I want the mingw64 stuff to be used first.

If the Rust installer for windows ever defaults to the MSVC version all of this will change! :) 

Personally I'd have preferred rusqlite just used the sqlite amalgam and maybe that'll be the path of most awesome if Rust on windows ever requires MSVC by default instead of MingW.

**Comment by gwenn** [comment]
It may interest you to know that there is an appveyor script that only uses:
 - rust stable (gnu version)
 - mingw
 - official pre-compiled dll

See https://github.com/jgallagher/rusqlite/pull/121

**Comment by photex** [comment]
@gwenn, a dll is undesirable for almost any use of sqlite that I have personally. I'd still cast the vote to fetch the amalgam instead.
For others that might be fine tho. I suppose it could be a cargo feature. ?

**Comment by jgallagher** [comment]
This is now available as the `bundled` feature (see #176 / #197).
